util/topsql/reporter/reporter.go

```diff
@@ -148,12 +155,16 @@ type SQLMeta struct {
 //
 // planBinaryDecoder is a decoding function which will be called asynchronously to decode the plan binary to string
 // MaxStatementsNum is the maximum SQL and plan number, which will restrict the memory usage of the internal LFU cache
-func NewRemoteTopSQLReporter(dataSink DataSink, decodePlan planBinaryDecodeFunc) *RemoteTopSQLReporter {
+func NewRemoteTopSQLReporter(decodePlan planBinaryDecodeFunc) *RemoteTopSQLReporter {
 	ctx, cancel := context.WithCancel(context.Background())
 	tsr := &RemoteTopSQLReporter{
-		ctx:                     ctx,
-		cancel:                  cancel,
-		dataSink:                dataSink,
+		ctx:    ctx,
+		cancel: cancel,
+
+		// An unbuffered channel for registration. A call for Register will block
+		// until the registration is successful or the RemoteTopSQLReporter is closed.
+		dataSinkRegCh: make(chan DataSink),
```

Comment:
It may be blocked for a long while since there is a 100ms in the loop and the reporter will not handle multiple register for once. Is this expected? (You will not face such long block when using mutex).

Reply:
I am adding context as a parameter. It looks better.

Reply:
It's by design. If the channel is buffered and the reporter is down, it provides the possibility that the datasink will not receive any feedback, and it will never know the reporter is down.